pkg/kncloudevents/message_dispatcher.go

```diff
@@ -0,0 +1,437 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kncloudevents
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/event"
+	cehttp "github.com/cloudevents/sdk-go/v2/protocol/http"
+	"github.com/hashicorp/go-retryablehttp"
+	"go.opencensus.io/trace"
+
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/pkg/network"
+	"knative.dev/pkg/system"
+
+	eventingapis "knative.dev/eventing/pkg/apis"
+
+	"knative.dev/eventing/pkg/broker"
+	"knative.dev/eventing/pkg/channel/attributes"
+	"knative.dev/eventing/pkg/tracing"
+	"knative.dev/eventing/pkg/utils"
+)
+
+const (
+	// noDuration signals that the dispatch step hasn't started
+	NoDuration = -1
+	NoResponse = -1
+)
+
+type DispatchInfo struct {
+	Duration       time.Duration
+	ResponseCode   int
+	ResponseHeader http.Header
+	ResponseBody   []byte
+}
+
+type SendOption func(*senderConfig) error
+
+func WithReply(reply *duckv1.Addressable) SendOption {
+	return func(sc *senderConfig) error {
+		sc.reply = reply
+
+		return nil
+	}
+}
+
+func WithDeadLetterSink(dls *duckv1.Addressable) SendOption {
+	return func(sc *senderConfig) error {
+		sc.deadLetterSink = dls
+
+		return nil
+	}
+}
+
+func WithRetryConfig(retryConfig *RetryConfig) SendOption {
+	return func(sc *senderConfig) error {
+		sc.retryConfig = retryConfig
+
+		return nil
+	}
+}
+
+func WithHeader(header http.Header) SendOption {
+	return func(sc *senderConfig) error {
+		sc.additionalHeaders = header
+
+		return nil
+	}
+}
+
+func WithTransformers(transformers ...binding.Transformer) SendOption {
+	return func(sc *senderConfig) error {
+		sc.transformers = transformers
+
+		return nil
+	}
+}
+
+type senderConfig struct {
+	reply             *duckv1.Addressable
+	deadLetterSink    *duckv1.Addressable
+	additionalHeaders http.Header
+	retryConfig       *RetryConfig
+	transformers      binding.Transformers
+}
+
+func SendEvent(ctx context.Context, event event.Event, destination duckv1.Addressable, options ...SendOption) (*DispatchInfo, error) {
+	message := binding.ToMessage(&event)
+
+	return SendMessage(ctx, message, destination, options...)
+}
+
+func SendMessage(ctx context.Context, message binding.Message, destination duckv1.Addressable, options ...SendOption) (*DispatchInfo, error) {
```

Comment:
@pierDipi even this is still WIP, this would be the signature for the function I had in mind. Any objections so far?

Reply:
One note, afaik, `Message` is not a buffered data structure meaning that consuming or transforming it to event twice might have an undefined behavior (aka it depends on the implementation), so unless `Message` is strictly required, why not accepting `Event`?

Reply:
We provide a `SendEvent` function too, which accepts an Event (some lines above). The `SendMessage` is mostly provided, because the existing API uses a `Message` too.